File: coding/src/com/algos/core/li17_interval/revised/NMeetingsInOneRoom.java
package  com.algos.core.li17_interval.revised;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import com.algos.core.models.Interval;

public class NMeetingsInOneRoom {
    public static void main(String args[]) {
        NMeetingsInOneRoom obj = new NMeetingsInOneRoom();

        int n = 6;
        int start[] = {1,3,0,5,8,5};
        int end[] = {2,4,5,7,9,9};
        System.out.println(obj.maxMeetings(start,end,n));
    }

    // Problem: https://practice.geeksforgeeks.org/problems/n-meetings-in-one-room-1587115620/1
    // Idea: You need to sort them on ending. On begenning won't work as one event starting at the begenning might space complete length.
    // Picking that allow us to pick more events. Hence, we need to pick events which ends quickly to maximize.
    // Activity selection is similar to this problem
    // Time: O(nlogn)
    public List<Integer> maxMeetings(int start[], int end[], int n) {
        List<Interval> intervals = new ArrayList<>();
        for(int i = 0; i < n; i++)
            intervals.add(new Interval(start[i], end[i], i+1));

        Collections.sort(intervals, (m1, m2) -> m1.end-m2.end);

        List<Integer> ans = new ArrayList<>();
        ans.add(intervals.get(0).pos);

        int lastEnd = intervals.get(0).end;
        for(int i = 1;i<n;i++) {
            if(intervals.get(i).start > lastEnd) {
                ans.add(intervals.get(i).pos);
                lastEnd = intervals.get(i).end;
            }
        }

        return ans;
    }
}

File: dsa/topics/17_intervals/python/nmeetings_in_one_room.py
# Problem: https://practice.geeksforgeeks.org/problems/n-meetings-in-one-room-1587115620/1
# Idea:
# Sort meetings based on end time.
# Picking the meeting that ends earliest leaves maximum room for remaining meetings.
# This is the classic Activity Selection Problem.
# Time: O(n log n)

def max_meetings(start, end, n):
    # create list of (end, start, position)
    intervals = []
    for i in range(n):
        intervals.append((end[i], start[i], i + 1))

    # sort by end time
    intervals.sort()

    ans = []
    ans.append(intervals[0][2])   # first meeting position
    last_end = intervals[0][0]

    for i in range(1, n):
        if intervals[i][1] > last_end:
            ans.append(intervals[i][2])
            last_end = intervals[i][0]

    return ans


# ------------------ Test (Same as Java main) ------------------

n = 6
start = [1, 3, 0, 5, 8, 5]
end = [2, 4, 5, 7, 9, 9]

print(max_meetings(start, end, n))
